modules/Bio/EnsEMBL/Hive/Accumulator.pm

```diff
@@ -131,6 +133,60 @@ sub dataflow {
     }
 }
 
+sub checkEmptyKeys {
+  my ( $self, $key_signature, $accu_address ) = @_;
+  # we get number of brackets, that are empty for key signature and address
+  # and verify that each empty brackets in key_signature is empty in adress
+  my @curvedBrackets = $self->findEmptyBrackets($key_signature, '{', '}');
+  my @curvedBracketsAddress = $self->findEmptyBrackets($accu_address, '{', '}');
+  my %addressBrackets = map { $_ => 1 } @curvedBracketsAddress;
+
+  foreach my $position ( @curvedBrackets ) {
+    if (!exists($addressBrackets{$position})) {
+        die "Null hash key in accumulator";
+    }
+  }
+
+  my @squareBrackets = $self->findEmptyBrackets($key_signature, '[', ']');
+  my @squareBracketsAddress = $self->findEmptyBrackets($accu_address, '[', ']');
+  %addressBrackets = map { $_ => 1 } @squareBracketsAddress;
+
+  foreach my $position ( @squareBrackets ) {
+    if (!exists($addressBrackets{$position})) {
+        die "Null hash key in accumulator";
```

Comment:
This one (for `[]`) should read Null array key. Since you have the positions of the array or hash elements that didn't get substituted, it would be kind to the user to include the original accu_address and the positions that caused the problem in the output message.

Otherwise, it seems like this works. I suspect that it would also be possible to use pattern matching to check this, but your approach has an advantage in readability and maintainability.

Reply:
@ens-bwalts all done, thanks for review